### ISSUE TITLE
Change format of lintserver return messages

### ISSUE THREAD

**Issue by davidanthoff** [tonyhffong/Lint.jl#131]
Could the format of the returned messages from the lint server be changed so that it is easier to parse?

Right now the format is:

```
filename:linenumber errornumber errormessage
```

So the different parts are separated either by `:` or just by a space. But both of these characters are also valid characters in filename and errormessage... Things would be much simpler to parse in a client if the format was e.g.

```
filename;linenumber;errornumber;errormessage
```

I could then just split the string by the first three `;` that appear in the messages into its parts.

**Comment by TotalVerb** [comment]
`;` is also a valid character in filename and errormessage.

**Comment by davidanthoff** [comment]
Oh, then just some character that is not valid in filenames. I'm less concerned about the errormessage, given that that is the last part.

**Comment by TotalVerb** [comment]
On Unix, the only characters not valid in filenames are `/` and null. I think a JSON response format would be more robust. I would be willing to implement that shortly.

**Comment by davidanthoff** [comment]
Yeah, JSON would definitely be even better. And maybe make the whole protocol a bit more robust by sending a header with something like
````
Content-length: 334\n
\n
````
Followed by the actual JSON. The content length number would correspond to the # of bytes used by the JSON of the actual message. This kind of header would be sent both for the request and the response.

This is loosely inspired by [this](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#base-protocol).

**Comment by TotalVerb** [comment]
I like that idea. If we are going to do that, we may as well implement the entire protocol there. However it seems that might be some more work than just JSON serialization.

I think `lintserver` needs to grow a `protocol=` keyword argument to avoid breaking existing code, so maybe we can add the full protocol later, and implement a raw JSON protocol for now?

**Comment by davidanthoff** [comment]
@TotalVerb I think that is the way to go, a raw JSON protocol for now sounds great.

I have some very crude (and probably buggy) response parsing right now in JuliaEditorSupport/julia-vscode#5. For now I'll just leave that as is and wait until you have added some JSON protocoll, instead of making my code more robust, ok?

**Comment by TotalVerb** [comment]
Yeah, that sounds like a good plan. Any change will preserve backwards compatibility.

**Comment by davidanthoff** [comment]
From my point of view this is now a lot less urgent. I came up with a pretty robust parsing for the VS Code extension, and I actually like that I don't need to take on more dependencies like JSON.jl...

**Comment by TeroFrondelius** [comment]
I made linter-julia package for Atom and I'm also interested about this.

**Comment by TotalVerb** [comment]
Thanks for the bump. I will want to get #172 and #173 ready and in first, and then I can take some time to look at this.

**Comment by TeroFrondelius** [comment]
Hi,
I would prefer atom-linter standard message:
```js
{
  severity: 'info',
  location: {
    file: name_of_the_file.jl,
    position: [[row-1, 0], [row-1, 1]],
    },
  excerpt: "I371",
  description: "use of undeclared symbol"
}
```
https://github.com/steelbrain/linter/blob/master/docs/examples/standard-linter-v2.md

This is version 2, which is not in use yet, but under development.

**Comment by davidanthoff** [comment]
There is also the Microsoft Language Server Protocol standard [here](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#diagnostic). I think that is implemented by more clients than the Atom (at least [this](https://github.com/Microsoft/language-server-protocol/wiki/Protocol-Implementations#editors-ides-supporting-the-protocol) list suggests so).

**Comment by TotalVerb** [comment]
There is no reason we cannot support both with keyword arguments.

**Comment by TeroFrondelius** [comment]
I was thinking something like this:
```julia
function convertmsgtojson(msg)
    evar = msg.variable
    txt = msg.message
    file = msg.file
    linenumber = msg.line
    errorrange = "[[$linenumber, 0],[$linenumber, 80]]"
    code = string(msg.code)
    if code[1] == "I"
        etype = "Info"
    elseif code[1] == "W"
        etype = "Warning"
    else
        etype = "Error"
    end

    return """{"type": $etype
               "text": $code $evar $txt
               "range": $errorrange,
               "filePath": $file}"""
end
```
This now just the linter v1 message style, but liked said earlier, others could be added. What do you think about this approach?

**Comment by davidanthoff** [comment]
Well, if we use JSON we need to properly encode things, i.e. the code above doesn't do that. I would strongly suggest to use the JSON.jl package to create the JSON message for that, it should handle all of that properly.

**Comment by davidanthoff** [comment]
Oh, and it might make sense to use a proper URI schema to encode the filename?

**Comment by TeroFrondelius** [comment]
So the first change would be:
```julia
function convertmsgtojson(msg)
    evar = msg.variable
    txt = msg.message
    file = msg.file
    linenumber = msg.line
    errorrange = "[[$linenumber, 0],[$linenumber, 80]]"
    code = string(msg.code)
    if code[1] == "I"
        etype = "Info"
    elseif code[1] == "W"
        etype = "Warning"
    else
        etype = "Error"
    end

    return json("""{"type": $etype
               "text": $code $evar $txt
               "range": $errorrange,
               "filePath": $file}""")
end
```
Or have I totally misunderstood? Can you give a link or code example of the URI schema?

**Comment by TotalVerb** [comment]
No, that's not correct. One should do

```
JSON.json(Dict(
    "type" => etype,
    "text" => "$code $evar $txt",
    "range" => errorange,
    "filePath" => file))
```

**Comment by TeroFrondelius** [comment]
Hi again. [I made an initial version](https://github.com/TeroFrondelius/Lint.jl/blob/42421434dfcb560e02f81c78f23c4bcf0125f2d5/src/Lint.jl#L359-L444). Tests and URI schema missing, but I would like to hear your comments.

**Comment by TotalVerb** [comment]
This seems incorrect

```julia
        loc = JSON.json(Dict("file" => file,"position" => errorrange))
        return JSON.json(Dict("severity" => etypenumber,
                              "location" => loc,
                              "excerpt" => code,
"description" => "$evar $txt"))
```

I believe that you mean to do
```julia
        return JSON.json(Dict("severity" => etypenumber,
                              "location" => Dict("file" => file,"position" => errorrange),
                              "excerpt" => code,
"description" => "$evar $txt"))
```

**Comment by TotalVerb** [comment]
@TeroFrondelius Thanks for your work on this. Could you open a pull request? it will be easier to review.

**Comment by TeroFrondelius** [comment]
@TotalVerb thanks for your review. I will open a pull request after I will add some basic tests. Currently all old tests pass. 

My main issue is to find how to transfer the msg.variable information to linter-julia (and how to use it there). Here is the linter type defined: https://github.com/steelbrain/linter/blob/master/lib/types.js. The easy solution is to embed the variable to text like it is currently done, but I would like to keep the window open and have it separately available. In my mind the whole point of the JSON is to be able to transfer the LintMessage type to the dark side.

**Comment by TotalVerb** [comment]
This was fixed by #194.
